[2기 류하준] 🚀 나만의 유튜브 강의실 미션 2단계

HaJunRyu opened the pull request:
# 🚀 나만의 유튜브 강의실 미션 2단계
## 🎯 요구사항

### 🚏 1) 라우팅 기능
* [x]  **Browser History Api**를 이용하여 SPA처럼 라우팅을 적용한다.
* [x]  routing은 `볼 영상(to-watch)`, `본 영상(watched)`, `좋아요 한 영상(liked)` 로 관리한다.

### 📚 2) 강의실 관리
* [x]  Browser History Api를 이용하여 SPA처럼 라우팅을 적용한다.
* [x]  로컬스토리지를 사용하여, 새로고침을 해도 데이터의 변경사항이 그대로 적용되야 한다.
* [x]  가장 처음에는 저장된 영상이 없음으로, 비어있다는 것을 사용자에게 알려주는 상태를 보여준다.
* [x]  이후 페이지를 방문했을 때 기본 메인 화면은 내가 볼 영상들의 리스트를 보여준다.
- 영상 카드의 이모지 버튼을 클릭하여 아래와 같은 상태 변경이 가능해야 한다.
  - * [x]  ✅ 본 영상으로 체크
  - * [x]  🗑️ 버튼으로 저장된 리스트에서 삭제할 수 있다. (삭제 시 사용자에게 정말 삭제할 것인지 물어봅니다.)
* [x]  본 영상, 볼 영상 버튼을 눌러 필터링 할 수 있다.
* [x]  👍 좋아요 버튼을 누른 데이터만 필터링해서 보여줄 수 있는 메뉴를 만든다.
* [x]  👍 좋아요 버튼을 누른 경우, 로컬에서 데이터를 변경한다.
* [x]  👍 좋아요 버튼을 다시 클릭해서 해지할 수 있어야 한다.

### 👋🏼  3)사용자 경험
* [x] 사용자가 버튼을 클릭했을 때 해당 행위가 정상적으로 동작하거나, 실패하였음을 snackbar를 통해 보여준다.

## 소감
초반에 나름(?) 구조를 잘 잡아놓은거 같아서 코드의 양이 많아져도 복잡도가 그렇게 높아지지 않아서 이번 미션은 1단계에 비해 많이 수월했습니다!
좀 아쉬웠던점은 diff알고리즘을 적용해도 list렌더링 시 앞쪽의 컴포넌트를 삭제하면 그 뒤의 컴포넌트들도 리렌더링이 된다는것이었는데 이건 react의 key props의 개념을 적용해보면 어떨까싶습니다.
확실히 ts를 쓰니 코드의 양이 많아질수록 (애플리케이션 규모가 늘어날수록) 제 힘을 발휘한다는 느낌을 받았습니다. 만약 ts를 안썻으면 어쩔뻔했을까... 이런 생각이 들었던것 같습니다.
아쉬운점은 리팩토링을 많이는 하지 못했다는점입니다. redux를 모방하여 reducer함수를 작성하고 있는데 확실히 reducer에서는 case마다 거의 비슷한 패턴을 사용하다보니 아무래도 중복코드가 엄청 많은데 이건 3단계에서 유틸함수를 만들어 리팩토링 해볼 예정입니다.

그리고 createStore함수처럼 router도 함수형으로 작성해보려고 코드를 완성시키고 작동도 시켯는데 기존의 store의 state를 기반으로 렌더링이 되는 컴포넌트들을 비교하는 applyDiff를 통해서 페이지 컴포넌트를 교체해주지 않으니 라우트가 바뀔때 a to z까지 다시 리렌더링 되는 이슈가 있어 Route컴포넌를 만들어 App컴포넌트의 children으로 렌더링 해주는 방식을 사용하여 해결했지만 처음의 방식보다 코드가 썩 마음에 들지는 않는것 같습니다ㅠ

+ 추가: 온라인세션 10분전에 안건데 Browser History로 구현해놓고 popState를 구현을 안했네요 하하,,, 3단계하면서 추가 해볼게요우~

imakerjun committed:
feat: init project

HaJunRyu committed:
Chore: parcel bundler로 프로젝트 모듈, 번들링 초기 세팅

+ parcel에서 async, await를 사용하기 위해 babel plugin 추가
+ ts 사용을 위한 index.js -> ts로 변경

new file:   .babelrc
modified:   index.html
new file:   package-lock.json
new file:   package.json
modified:   src/css/shared/index.css
modified:   src/css/ui/index.css
renamed:    src/js/index.js -> src/js/index.ts

HaJunRyu committed:
Chore: js폴더명 -> ts로 변경

renamed:    src/js/index.ts -> src/ts/index.ts

HaJunRyu committed:
Feat: Axios 의존성 모듈 추가 및 YoutubeAPI 연결

modified:   package-lock.json
modified:   package.json
new file:   src/ts/api/customAxios.ts
new file:   src/ts/api/youtube.ts

HaJunRyu committed:
Feat: 컴포넌트를 생성하고 마운트시키는 함수 정의

new file:   src/ts/domHelper/createNode.ts

HaJunRyu committed:
Feat: Component 타입 정의 및 types 진입점 파일로 모듈화

new file:   src/ts/types/Component.ts
new file:   src/ts/types/index.ts

HaJunRyu committed:
Feat: ceateNode함수 사용법 변경

children의 타입이 Node[]였는데 이렇게 하면 사용할때 배열내부에서 컴포넌트 함수를 호출해야하는 일이 발생하여
createNode 함수내부에서 children요소들을 호출할 수 있게끔 변경

modified:   src/ts/domHelper/createNode.ts

HaJunRyu committed:
Feat: 애플리케이션 CSR화 (아주 간략하게만 계속 세분화 할것임)

html 파일의 태그들을 template삼아 크게 Home페이지와 Modal컴포넌트로 나누어 CSR완료

컴포넌트들의 단위는 UI의 가장 최상위 컴포넌트인 app, 페이지 컴포넌트인 pages, 일반적인 컴포넌트 요소들인 components 폴더로 나누어 생성

Changes to be committed:
modified:   index.html
new file:   src/ts/app/App.ts
new file:   src/ts/components/Modal.ts
new file:   src/ts/components/index.ts
new file:   src/ts/domHelper/index.ts
modified:   src/ts/index.ts
new file:   src/ts/pages/Home.ts
new file:   src/ts/pages/index.ts

HaJunRyu committed:
Fix: App컴포넌트에 id="app"이 빠져 스타일링 미적용되는 버그 수정

Changes to be committed:
modified:   src/ts/app/App.ts

HaJunRyu committed:
Feat: 전역 상태관리를 위한 store로직 및 타입 추가

new file:   src/ts/store/actionCreator.ts
new file:   src/ts/store/actionType.ts
new file:   src/ts/store/createStore.ts
new file:   src/ts/store/index.ts
new file:   src/ts/store/reducer.ts
new file:   src/ts/types/Store.ts
modified:   src/ts/types/index.ts

HaJunRyu committed:
Feat: 프로젝트 구조 수정

컴포넌트 타입과 createNode의 반환값을 Element로 변경, 컴포넌트의 props를 필수값으로 변경

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/domHelper/createNode.ts
modified:   src/ts/index.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/types/Component.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Feat: 버튼 컴포넌트 생성

Changes to be committed:
new file:   src/ts/components/Button.ts

HaJunRyu committed:
Feat: Heading 컴포넌트 생성

Changes to be committed:
new file:   src/ts/components/Heading.ts

HaJunRyu committed:
Feat: Navigator 컴포넌트 생성

new file:   src/ts/components/Navigator.ts

HaJunRyu committed:
Feat: Header 컴포넌트 생성

Changes to be committed:
new file:   src/ts/components/Header.ts

HaJunRyu committed:
Feat: Video 컴포넌트 생성

가장 많이 재활용 될 컴포넌트로 예상

Changes to be committed:
new file:   src/ts/components/Video.ts

HaJunRyu committed:
Feat: Video컴포넌트를 리스트처럼 담을 SaveVideoSection 컴포넌트 생성

Changes to be committed:
new file:   src/ts/components/SaveVideoSection.ts

HaJunRyu committed:
Feat: 전역상태를 가져와 saveVideoList를 렌더링해주는 Main컴포넌트 추가

Changes to be committed:
new file:   src/ts/components/Main.ts

HaJunRyu committed:
Feat: 현재까지 만든 컴포넌트를 기반으로 Home 컴포넌트에 렌더링

Changes to be committed:
modified:   src/ts/pages/Home.ts

HaJunRyu committed:
Feat: 현재까지 추가한 컴포넌트들을 index.ts파일에 모듈화

Changes to be committed:
modified:   src/ts/components/index.ts

HaJunRyu committed:
Feat: App컴포넌트에 Main영역과 Modal영역 렌더링

Changes to be committed:
modified:   src/ts/app/App.ts

HaJunRyu committed:
Feat: Modal 컴포넌트의 구성요소 CSR화 완료

+ youtube API에서 주는 데이터의 형식으로 비디오 타입 정의

Changes to be committed:
modified:   src/ts/components/Button.ts
new file:   src/ts/components/Form.ts
modified:   src/ts/components/Header.ts
new file:   src/ts/components/Input.ts
modified:   src/ts/components/Main.ts
modified:   src/ts/components/Modal.ts
new file:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/components/index.ts
modified:   src/ts/pages/Home.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Component.ts
modified:   src/ts/types/Store.ts
modified:   src/ts/types/index.ts

HaJunRyu committed:
Refactor: createNode 메서드가 children을 받을때 Falsy한값이 있다면 무시하게끔 수정

Changes to be committed:
modified:   src/ts/domHelper/createNode.ts

HaJunRyu committed:
Feat: 상태를 기반으로 모달을 Toggle하게끔 변경

index.ts에 상태가 변경되면 render함수가 호출되게끔 구독

Changes to be committed:
modified:   src/css/ui/modules/modal.css
modified:   src/ts/app/App.ts
modified:   src/ts/components/Button.ts
modified:   src/ts/components/Modal.ts
modified:   src/ts/index.ts
modified:   src/ts/pages/Home.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Feat: onsubmit이벤트로 유튜브 검색 기능 구현

input의 값을 상태로 관리하고 싶었으나 가상돔 구현 실패로 너무 비효율적인 동작과 버그를 방지하기 위해 상태로 사용하지 않았음

Changes to be committed:
modified:   src/css/ui/modules/modal.css
modified:   src/css/ui/modules/youtubeCard.css
modified:   src/ts/api/youtube.ts
modified:   src/ts/components/Form.ts
modified:   src/ts/components/Input.ts
modified:   src/ts/components/Main.ts
modified:   src/ts/components/Modal.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Feat: 동영상의 게시 날짜를 표시할 수 있음

Changes to be committed:
modified:   src/css/ui/modules/youtubeCard.css
modified:   src/ts/components/Main.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
new file:   src/ts/utils/dateFormMaker.ts
new file:   src/ts/utils/transferCreateData.ts

HaJunRyu committed:
Feat: 검색시 스켈레톤 UI 추가

Changes to be committed:
modified:   src/css/ui/index.css
new file:   src/css/ui/modules/skeleton.css
modified:   src/ts/components/Modal.ts
modified:   src/ts/components/SearchVideoSection.ts
new file:   src/ts/components/Skeleton.ts
modified:   src/ts/components/index.ts

HaJunRyu committed:
Feat: 검색결과가 없을시 검색결과가 없다는 이미지와 메세지 추가

Changes to be committed:
modified:   src/ts/components/Modal.ts
new file:   src/ts/components/NotFound.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/index.ts

HaJunRyu committed:
Feat: 유튜브 검색 기능에 무한스크롤 기능 추가

Changes to be committed:
modified:   src/ts/api/youtube.ts
modified:   src/ts/components/Modal.ts
new file:   src/ts/components/Observe.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Feat: 리프레시시에도 가장 마지막에 검색한 키워드의 검색결과를 볼 수 있게끔 변경

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Modify: actionCreator에서 localStorage에 접근하던것을 이벤트 리스너로 이동

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/store/actionCreator.ts

HaJunRyu committed:
Feat: 검색한 동영상 저장기능 추가(저장된 영상은 저장버튼이 보이지 않게끔)

Changes to be committed:
modified:   src/ts/components/Main.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Feat: 최근 검색 키워드 저장 기능 추가

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Docs: README update

Changes to be committed:
modified:   README.md

HaJunRyu committed:
Refactor: createStore에서 선언된 초기상태값을 Reducer함수가 선언된 파일로 이동

Changes to be committed:
modified:   src/ts/store/createStore.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Chore: gitignore에 .DS_Store 불필요한 파일 목록 추가

Changes to be committed:
modified:   .gitignore

HaJunRyu committed:
Style: 사용하지 않는 console.log 제거

Changes to be committed:
modified:   src/ts/components/Video.ts

HaJunRyu committed:
Fix: 초기 렌더링시 Action이 객체라는걸 인지하지 못해 Action.type을 참조하는 라인에서 발생하는 버그 해결

Changes to be committed:
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Feat: diff알고리즘 적용

변경되지 않은 요소들은 리렌더링하지 않게끔 비교 알고리즘 추가 및 적용

Changes to be committed:
new file:   src/ts/domHelper/applyDiff.ts
modified:   src/ts/domHelper/index.ts
new file:   src/ts/domHelper/renderRoot.ts
modified:   src/ts/index.ts

HaJunRyu committed:
Feat: 저장된 영상이 없을때 띄워주는 메시지 추가

Changes to be committed:
modified:   src/css/ui/index.css
new file:   src/css/ui/modules/emptyMessage.css
new file:   src/ts/components/EmptyMessage.ts
modified:   src/ts/components/Main.ts
modified:   src/ts/components/index.ts

HaJunRyu committed:
Fix: 무한스크롤시 검색키워드가 유지되지 않던 버그 수정

+ 최근 검색 키워드 기록이 중복으로 되지 않게끔 수정

Changes to be committed:
modified:   src/ts/components/Modal.ts

HaJunRyu committed:
Fix: 새로운 검색결과에 저장버튼이 업데이트 되지 않는 버그 수정

+ 마지막 검색키워드와 nextPageToken값을 로컬스토리지에 저장하여 새로고침해도 무한스크롤시 검색결과가 정상적으로 나오게 수정

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Feat: YoutubeVideo 상태에 isLike, isWatched 추가

본 영상과 좋아하는 영상 섹션에서 나누어주려는 목적

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/pages/Home.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Refactor: 라우트 적용을 위한 페이지 컴포넌트 구조 개선

- Home페이지를 삭제하며 네비게이터는 App컴포넌트로 이동
- ToWatch 페이지 컴포넌트를 새로 생성, root경로에 매칭하여 사용 할 계획

Changes to be committed:
modified:   src/ts/app/App.ts
modified:   src/ts/components/index.ts
deleted:    src/ts/pages/Home.ts
new file:   src/ts/pages/Like.ts
renamed:    src/ts/components/Main.ts -> src/ts/pages/ToWatch.ts
new file:   src/ts/pages/Watched.ts
modified:   src/ts/pages/index.ts

HaJunRyu committed:
Refactor: Navigator 컴포넌트 추상화

Changes to be committed:
modified:   src/ts/app/App.ts
modified:   src/ts/components/Navigator.ts

HaJunRyu committed:
Refactor: 로컬 스토리지의 키를 상수값으로 관리 및 사용하게끔 변경

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/store/reducer.ts
new file:   src/ts/utils/localStorageKey.ts

HaJunRyu committed:
Feat: 저장된 비디오를 삭제하는 기능 추가

Changes to be committed:
modified:   src/ts/components/SaveVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/index.ts

HaJunRyu committed:
Feat: 저장된 비디오 삭제시 정말 삭제할거냐는 confirm 메시지 추가

Changes to be committed:
modified:   src/ts/components/SaveVideoSection.ts

HaJunRyu committed:
Feat: route 모듈 정의 및 적용

route컴포넌트를 하나 만들어 패스가 변경될떄마다 해당 컴포넌트의 children들을 교체해주는 방식

Changes to be committed:
modified:   src/ts/app/App.ts
new file:   src/ts/components/Link.ts
modified:   src/ts/components/Navigator.ts
modified:   src/ts/components/index.ts
modified:   src/ts/pages/Like.ts
modified:   src/ts/pages/Watched.ts
modified:   src/ts/pages/index.ts
new file:   src/ts/route/createRouter.ts
new file:   src/ts/route/index.ts
new file:   src/ts/route/path.ts

HaJunRyu committed:
Modify: import 순서를 store를 최우선으로 변경

Changes to be committed:
modified:   src/ts/components/Modal.ts
modified:   src/ts/pages/ToWatch.ts

HaJunRyu committed:
Refactor: diff 알고리즘이 적용되게끔 라우터 로직 변경

path를 바로 참조하여 독립적으로 페이지를 그려주는것이 아닌 path를 상태로 관리하여 상태에 따라 페이지를 렌더링하게끔 변경

Changes to be committed:
modified:   src/ts/app/App.ts
modified:   src/ts/components/Navigator.ts
new file:   src/ts/pages/NotFound.ts
modified:   src/ts/pages/index.ts
modified:   src/ts/route/createRouter.ts
modified:   src/ts/route/index.ts
modified:   src/ts/route/path.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/reducer.ts
new file:   src/ts/types/Route.ts
modified:   src/ts/types/Store.ts
modified:   src/ts/utils/localStorageKey.ts

HaJunRyu committed:
Modify: Youtube API search 요청시 video데이터만 검색되게끔 파라미터 추가

Changes to be committed:
modified:   src/ts/api/youtube.ts

HaJunRyu committed:
Feat: 본 영상으로 추가하는 기능 (isWatched 상태 dispatch) 추가

Changes to be committed:
modified:   src/css/ui/modules/youtubeCard.css
modified:   src/ts/components/SaveVideoSection.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/pages/ToWatch.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Feat: 액션이 일어났을때 사용자에게 알려주기 위한 Snackbar 컴포넌트 추가

Changes to be committed:
modified:   src/css/ui/index.css
new file:   src/css/ui/modules/snackbar.css
modified:   src/ts/app/App.ts
modified:   src/ts/components/SaveVideoSection.ts
new file:   src/ts/components/Snackbar.ts
modified:   src/ts/components/index.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/reducer.ts
modified:   src/ts/types/Store.ts

HaJunRyu committed:
Feat: 볼 영상 라우트에서 볼 영상들의 리스트만 렌더링하는 기능 추가

Changes to be committed:
modified:   src/ts/pages/ToWatch.ts

HaJunRyu committed:
Feat: 유튜브 비디오 리스트가 isWatched에 따라 각각의 라우트에 렌더링되는 기능 추가

Changes to be committed:
modified:   src/ts/pages/ToWatch.ts
modified:   src/ts/pages/Watched.ts

HaJunRyu committed:
Feat: 라우트에 따라 네비게이터 버튼을 하이라이트 하는 기능 추가

Changes to be committed:
modified:   src/ts/components/Navigator.ts

HaJunRyu committed:
Refactor: 볼, 본 영상으로 토클할 때 snackbar에 각각의 메세지를 띄워주게끔 변경

Changes to be committed:
modified:   src/ts/components/SaveVideoSection.ts

HaJunRyu committed:
Style: 사용하지 않는 console.log 함수 사용 코드 제거

Changes to be committed:
modified:   src/ts/components/Navigator.ts

HaJunRyu committed:
Feat: 좋아요 버튼으로 isSave상태를 변경해 좋아하는 영상 라우트에서 렌더링하는 기능 추가

Changes to be committed:
modified:   src/css/index.css
modified:   src/css/ui/modules/youtubeCard.css
modified:   src/ts/components/SaveVideoSection.ts
modified:   src/ts/components/SearchVideoSection.ts
modified:   src/ts/components/Video.ts
modified:   src/ts/pages/Like.ts
modified:   src/ts/pages/ToWatch.ts
modified:   src/ts/pages/Watched.ts
modified:   src/ts/store/actionCreator.ts
modified:   src/ts/store/actionType.ts
modified:   src/ts/store/reducer.ts

HaJunRyu committed:
Merge: origin 저장소의 hajunryu-step2 브랜치 pull 후 난 conflict 해결

Merge branch 'hajunryu-step2' of https://github.com/next-step/js-youtube-classroom into hajunryu-step2

Conflicts:
README.md
index.html
src/css/index.css
src/css/shared/index.css
src/css/ui/index.css
src/css/ui/modules/modal.css
src/css/ui/modules/youtubeCard.css

Changes to be committed:
modified:   README.md

HaJunRyu committed:
Refactor: video의 상태를 바꿔주는 이벤트 핸들러들의 중복을 제거, snackbar UI에 debounce 적용

Changes to be committed:
modified:   src/ts/components/SaveVideoSection.ts
new file:   src/ts/utils/debounce.ts

younoah reviewed (commented):
안녕하세요 하준님 리뷰를 하게된 정윤호라고합니다.
하준님의 코드를 바이블 삼아 뜯어보면서 학습해도 될 것 같다는 생각이 들정도로 엄청난 것을 본 것 같습니다.
아직 실력이 많이 모자라서 감히 제가 피드백을 작성할 수가 없네요. 죄송해요🤣
리액트와 타입스크립트 학습하면서 다시 레벨3스터디를 할 때 하준님 코드를 참고하면 큰 도움이 될 것 같아요!
고생하셨고 남은 기간동안 화이팅입니다!

inhyuck commented on index.html:
```diff
@@ -1,124 +1,15 @@
-    </div>
-    <script type="module" src="./src/js/index.js"></script>
+    <noscript
+      >이 사이트를 이용하려면? 사용 중인 웹 브라우저에서 JavaScript를 활성화 해야 합니다.</noscript
```
친절한 설명 좋네요 ㅎㅎ

inhyuck commented on src/ts/types/Store.ts:
```diff
@@ -0,0 +1,70 @@
+  error: Error;
+  searchList: YoutubeVideo[];
+  recentSearchKeywords: string[];
+  saveVideoList: YoutubeVideo[];
```
ts 라 타입이 명시되어 크게 상관없을수도 있겠지만 변수네이밍만 봤을때는 비디오리스트를 저장하는 함수명으로 착각할 수도 있을것같아요. savedVideoList 와 같이 명사형은 어떨까요?

HaJunRyu replied:
좋은 지적 감사합니다! save가 동사이다보니 함수라는 느낌도 드네요 정말 🥲

inhyuck commented on src/ts/store/createStore.ts:
```diff
@@ -0,0 +1,39 @@
+  const subscribe = (newListener: Function) => {
+    listeners.push(newListener);
+
+    return () => {
+      listeners = listeners.filter(listener => listener !== newListener);
+    };
```
unSubscribe() 함수를 반환하는게 맞을까요? 디테일한 구현이 멋지네요 👍

HaJunRyu replied:
맞습니다! subscribe를 호출할때 반환값을 받아뒀다가 적절한때 구독해제를 할 수 있는 함수입니다ㅎㅎ

inhyuck commented on src/ts/components/NotFound.ts:
```diff
@@ -0,0 +1,18 @@
+interface Props extends CommonProps {}
+
+const NotFound: Component<Props> = () => {
+  const $notFount = createNode(`
```
사소한거지만... 오타인것같아요!

HaJunRyu replied:
ㅠㅠ...좀 더 천천히 꼼꼼히 치는 습관을 들여야겠네요...

inhyuck commented on src/ts/components/Navigator.ts:
```diff
@@ -0,0 +1,64 @@
+    dispatch(modalOpenAction());
+  };
+
+  const { navigate } = route;
```
navigate 함수가 onNavigatehandler 함수내에서 호출되므로 onNavigatehandler 함수 바로 위 라인으로 이동하는건 어떨까요? 가급적 해당 변수를 사용하는곳과 가까운게 가독성에 좋을 것 같습니다.

HaJunRyu replied:
좋은 의견 감사합니다! 저는 그냥 대부분 선언을 위쪽에 몰아버리는편인데 인혁님 말씀대로 그런 패턴을 사용하면 좀 더 가독성이 좋아지겠네요ㅎㅎ

inhyuck commented on src/ts/route/createRouter.ts:
```diff
@@ -0,0 +1,44 @@
+  };
+
+  const navigate = (path: string) => {
+    if (lastPath === path) return;
```
변경되는 라우트 path 가 이전과 동일해도 일반적으로 같은 메뉴버튼을 다시 클릭하면 refresh 하는 식의 동작을 기대하는 것 같아요. 해당 조건문에서는 이전과 라우트 path 가 동일하면 아무런 동작을 하지 않게 될 것 같습니다.

HaJunRyu replied:
넵 일부러 그 의도로 작성한거 맞습니다! 이미 해당 라우트를 보고 있는데 다시 라우트값을 바꾸어주어 vitualDOM을 생성해서 비교할 필요가 없다는 생각 때문이었습니다ㅎㅎ

inhyuck reviewed (commented):
안녕하세요 하준님. ts 로 짜여진 정교한 코드들 잘 봤습니다. 라우터부분도 실제 react-router 를 사용하는것과 거의 동일한 api를 제공하고 있는것 같고 상태관리쪽 코드도 멋지네요. 수고 많으셨습니다~ 👍

HaJunRyu committed:
Style: 사용하지 않는 import 구문 제거

Changes to be committed:
modified:   src/ts/components/Button.ts
modified:   src/ts/components/Header.ts
modified:   src/ts/domHelper/createNode.ts
modified:   src/ts/index.ts
modified:   src/ts/route/index.ts
modified:   src/ts/store/createStore.ts

HaJunRyu committed:
Style: 불필요한 주석 제거

Changes to be committed:
modified:   src/ts/app/App.ts

HaJunRyu committed:
Modify: 각각의 컴포넌트들이 전역 상태를 참조하기 때문에 render메서드에 state전달 값 삭제

Changes to be committed:
modified:   src/ts/store/createStore.ts

HaJunRyu committed:
Feat: createRouter에 pushState 이벤트 등록

Changes to be committed:
modified:   src/ts/route/createRouter.ts
